### ISSUE TITLE
yarn install does not work against external optional dependency

### ISSUE THREAD

**Issue by jeffrson** [yarnpkg/yarn#6665]
I'm using yarn 1.12.3. I need to install pm2 via local package registry that acts as a proxy for modules/package dependencies. The client has no internet connection itself.

The creators of pm2 decided to include a kind of usage counter - they have
```
 "optionalDependencies": {
    "gkt": "https://tgz.pm2.io/gkt-1.0.0.tgz"
  },
```
in their package.json. IMO this is wrong. Nevertheless, they refuse to change it because it's optional and yarn should ignore it just in case. However, yarn fails for this optional dependency if it can't reach the server tgz.pm2.io, even with "--ignore-optional". So it doesn't ignore it. Current stance seems to be, yarn needs to read the dependency in order to include it in yarn.lock no matter what (maybe someone might document what "optional" means exactly).

Unfortunately, me, as an ordinary mortal, am left alone with the conflict.

So I decided to give yarn its yarn.lock "preinitialized":
```
"gkt@https://tgz.pm2.io/gkt-1.0.0.tgz":
  version "1.0.0"
  resolved "https://tgz.pm2.io/gkt-1.0.0.tgz#405502b007f319c3f47175c4474527300f2ab5ad"
```

Now, with "--ignore-optional" yarn worked for me. That is, it did for 1.10.1:
```
>yarn --registry http://registry.local --ignore-optional
yarn install v1.10.1
[1/4] Resolving packages...
[2/4] Fetching packages...
error https://tgz.pm2.io/gkt-1.0.0.tgz: getaddrinfo ENOENT tgz.pm2.io tgz.pm2.io:443
info fsevents@1.2.4: The platform "win32" is incompatible with this module.
info "fsevents@1.2.4" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
[4/4] Building fresh packages...
success Saved lockfile.
Done in 5.72s.
```
Unfortunately, starting with 1.11.0 it does no longer install modules. And there's no error message:
```
>yarn --registry http://registry.local --ignore-optional
yarn install v1.12.3
[1/4] Resolving packages...
[2/4] Fetching packages...
[##########################################################] 263/264
```
Nothing is written to the current folder. Neither yarn install, nor another yarn add, do anything usable.
When I try to reproduce it here, it works as soon as there is internet available. But this "solution" is impossible for the targetted client.

IMO, optional dependencies should be ignored at all with "--ignore-optional". But at least yarn should work (again) with the dependency given in yarn.lock as it did before 1.11.0.

FYI - node 10.13.0, Windows 10 x64

**Comment by jeffrson** [comment]
I had tried with 1.12.3 before - it doesn't add this "integrity" line. I repeated my test with this line appended. yarn.lock actually looks like you wrote:
```
"gkt@https://tgz.pm2.io/gkt-1.0.0.tgz":
  version "1.0.0"
  resolved "https://tgz.pm2.io/gkt-1.0.0.tgz#405502b007f319c3f47175c4474527300f2ab5ad"
  integrity sha1-405502b007f319c3f47175c4474527300f2ab5ad
```
Unfortunately this didn't change anything :-(
```
>yarn --registry http://registry.local --ignore-optional
yarn install v1.12.3
[1/4] Resolving packages...
[2/4] Fetching packages...
[####################################################] 263/264
```

**Comment by jeffrson** [comment]
Not fixed with yarn 1.13.0 - behaviour not changed.

**Comment by arcanis** [comment]
> Any news?

Nope. Since `pm2` is the only package that's affected, and since they're basically abusing the system, this is very very low-priority (so low priority that I'm not even sure it *should* be fixed. We don't particularly want to encourage this kind of usage).

**Comment by jeffrson** [comment]
Well, it's not only that yarn handles optional dependencies in another manner than pm2 would want it to do.

There's something broken in handling network errors, as I wrote in
https://github.com/yarnpkg/yarn/issues/6665#issuecomment-440704585

**Comment by jeffrson** [comment]
AFAICT, the network issues are solved with 1.15.2.

Furthermore, as a reference, the pm2 dependency problem seems to be best worked around by appending this to package.json:
```
    "resolutions": {
        "gkt": "1.0.0"
    }
```

Hence, this report may be closed as far as I'm concerned.
